source/endpoints/ruleset.c

```diff
@@ -0,0 +1,878 @@
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(node);
+    AWS_PRECONDITION(table);
+
+    struct member_parser_wrapper wrapper = {
+        .allocator = allocator, .table = table, .all_members_parsed_successfully = true};
+
+    if (aws_json_const_iterate_object(node, member_fn, &wrapper)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Member field has unexpected type.");
+        return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+    }
+
+    if (!wrapper.all_members_parsed_successfully) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/*
+******************************
+* Parser functions.
+******************************
+*/
+
+static struct aws_endpoints_function *s_try_parse_function(
+    struct aws_allocator *allocator,
+    const struct aws_json_value *node);
+
+static struct aws_endpoints_reference *s_try_parse_reference(
+    struct aws_allocator *allocator,
+    const struct aws_json_value *node) {
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(node);
+
+    struct aws_json_value *ref_node = aws_json_value_get_from_object(node, aws_byte_cursor_from_c_str("ref"));
+    if (ref_node != NULL) {
+        struct aws_byte_cursor ref_cur;
+        aws_json_value_get_string(ref_node, &ref_cur);
+        return aws_endpoints_reference_new(allocator, aws_string_new_from_cursor(allocator, &ref_cur));
+    }
+
+    return NULL;
+}
+
+static struct aws_endpoints_expr *s_parse_expr(struct aws_allocator *allocator, const struct aws_json_value *node);
+
+static bool s_on_expr_element(size_t idx, const struct aws_json_value *value_node, void *user_data) {
+    (void)idx;
+    struct array_parser_wrapper *wrapper = user_data;
+    struct aws_endpoints_expr *expr = s_parse_expr(wrapper->allocator, value_node);
+
+    if (expr == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to parse value.");
+        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+        return false;
+    }
+
+    aws_array_list_push_back(wrapper->array, &expr);
+    return true;
+}
+
+static struct aws_endpoints_expr *s_parse_expr(struct aws_allocator *allocator, const struct aws_json_value *node) {
+    /* TODO: this recurses. in practical circumstances depth will never be high,
+    but we should still consider doing iterative approach */
+    if (aws_json_value_is_string(node)) {
+        struct aws_byte_cursor cur;
+        aws_json_value_get_string(node, &cur);
+        struct aws_endpoints_expr *expr = aws_endpoints_expr_new(allocator, AWS_ENDPOINTS_EXPR_STRING);
+        expr->e.string = aws_string_new_from_cursor(allocator, &cur);
+        return expr;
+    } else if (aws_json_value_is_number(node)) {
+        double number;
+        aws_json_value_get_number(node, &number);
+        struct aws_endpoints_expr *expr = aws_endpoints_expr_new(allocator, AWS_ENDPOINTS_EXPR_NUMBER);
+        expr->e.number = (int32_t)number;
```

Comment:
endpoints 2.0 spec does not say what the number is currently. the only place its used is substring function and that defines arg as int and other implementations use int32 as type for all numbers. Im following up on that. Will add comment to clarify that it might change.

Reply:
reverted this to be double in parsing and functions can cast it accordingly during eval